Modules/Test/classes/class.ilObjTestAccess.php

```diff
@@ -19,6 +19,109 @@
 */
 class ilObjTestAccess extends ilObjectAccess implements ilConditionHandling
 {
+    public function canBeDelivered(ilWACPath $ilWACPath)
+    {
+        $path_and_test = $this->pathAndTestId($ilWACPath->getPath());
+
+        if (!$path_and_test || !$path_and_test['test']) {
+            return false;
+        }
+
+        $object_id = (int) ilObjTest::_getObjectIDFromTestID($path_and_test['test']);
+        if (!$object_id) {
+            return false;
+        }
+
+        $references = ilObject2::_getAllReferences($object_id);
+
+        return $this->hasAccessTo('read', $references) && $this->roleBasedCheck($path_and_test['test'], $references, $path_and_test['path']);
+    }
+
+    private function isAnonymous() : bool
+    {
+        global $DIC;
+
+        return $DIC->user()->isAnonymous();
+    }
+
+    private function accessCodeOk(string $file, int $test_id) : bool
+    {
+        global $DIC;
+
+        $code = ilSession::get(ilTestSession::ACCESS_CODE_SESSION_INDEX)[$test_id] ?: false;
+
+        return $code && $this->isOwner($test_id, $file, $code);
+    }
+
+    private function isOwner(int $test_id, string $file, string $code = null) : bool
+    {
+        global $DIC;
+
+        $where = [
+            'user_fi = %s',
+            'value1 = %s',
+            'anonymous_id ' . (null === $code ? 'IS' : '=') . ' %s',
+            'test_fi = %s',
+        ];
+
+        $result = $DIC->database()->queryF(
+            'SELECT 1 FROM tst_solutions INNER JOIN tst_active ON active_id = active_fi WHERE ' . implode(' AND ', $where),
+            ['integer', 'text', 'text', 'integer'],
+            [$DIC->user()->getId(), $file, $code, $test_id]
+        );
+
+        return (bool) $DIC->database()->numRows($result);
+    }
+
+    /**
+     * @param int $test_id
+     * @param string[] $references
+     * @param string $file
+     *
+     * @return bool
+     */
+    private function roleBasedCheck(int $test_id, array $references, string $file) : bool
+    {
+        return ($this->isAnonymous() && $this->accessCodeOk($file, $test_id)) || $this->hasAccessTo('write', $references) || $this->isOwner($test_id, $file);
```

Comment:
Maybe you should use `\ilTestAccess::checkResultsAccessForActiveId` to check the access to results of a learner, **instead** of using asking `$this->hasAccessTo('write', $references)`.

Reply:
Do you mean instead of `$this->hasAccessTo('read', $references)`. The `write` check is to see if the current user is a teacher not a learner.

Reply:
No, I meant `$this->hasAccessTo('write', $references).`. As you wrote, this is the case to check if the results **of** a learner are accessible **to** someone with _higher_ permissions (reporting, results, etc.).